### DISCUSSION TITLE
Adding ability to parse C++ headers into GDT

### DISCUSSION THREAD

**Discussion by lightspot21** [NationalSecurityAgency/ghidra#6213] [Ideas]
I am currently looking into extending Ghidra's C parser to handle C++ headers. Specifically, I'd like to be able to create gdt files from Microsoft MFC headers, which are too big and complex to clean by hand.

Of course I'll be using a ready-made C++ parser, like the one included with Eclipse CDT (which already seems to be used in GhidraDev and is native Java, as suggested in https://stackoverflow.com/a/27496664), and at least for now, all I'll be doing is to extract automatically structs (class data, vtable structs, function definitions).

 I'd really like for this work, if at all possible, to be included in the upstream. Would the Ghidra team be opposed to a PR implementing such functionality?

Thanks in advance!

**Comment by Wall-AF** [comment]
Don't forget the `_based` (`__based`) datatypes too!

**Reply by lightspot21** [reply]
what do you mean?

**Reply by Wall-AF** [reply]
MSVC 1.5X uses definitions like:
```
D:\WIN311\MSVC151>findstr /s /n  _based *.java *.h *.hh *.hp* *.hx* *.inc *.inl *.c *.cc *.cp *cpp *.cx* *.asm *.xml  | more
INCLUDE\INITGUID.H:37:    EXTERN_C const GUID CDECL __based(__segname("_CODE")) name \
INCLUDE\MALLOC.H:20:#define __based     _based
INCLUDE\MALLOC.H:31:#define _NULLOFF    ((void __based(void) *)0xffff)
INCLUDE\MALLOC.H:76:void __based(void) * __cdecl _bcalloc(__segment, size_t, size_t);
INCLUDE\MALLOC.H:77:void __based(void) * __cdecl _bexpand(__segment,
INCLUDE\MALLOC.H:78:    void __based(void) *, size_t);
INCLUDE\MALLOC.H:79:void __cdecl _bfree(__segment, void __based(void) *);
INCLUDE\MALLOC.H:81:int __cdecl _bheapadd(__segment, void __based(void) *, size_t);
INCLUDE\MALLOC.H:87:void __based(void) * __cdecl _bmalloc(__segment, size_t);
INCLUDE\MALLOC.H:88:size_t __cdecl _bmsize(__segment, void __based(void) *);
INCLUDE\MALLOC.H:89:void __based(void) * __cdecl _brealloc(__segment,
INCLUDE\MALLOC.H:90:    void __based(void) *, size_t);
INCLUDE\NEW.H:19:#define _NULLOFF    ((void __based(void) *)0xffff)
INCLUDE\OLESTD.H:28:#define _based(a)
INCLUDE\OLESTD.H:407:           static char _based(_segname("_CODE")) _szAssertFile[]= __FILE__;
MFC\INCLUDE\AFXEXT.H:660:typedef char _based((_segment)_self) *BPSTR;
MFC\INCLUDE\AFXVER_.H:201:#define BASED_CODE __based(__segname("_CODE"))
MFC\INCLUDE\AFXVER_.H:205:#define BASED_DEBUG __based(__segname("AFX_DEBUG3_TEXT"))
MFC\INCLUDE\AFXVER_.H:209:#define BASED_STACK __based(__segname("_STACK"))
MFC\SAMPLES\DLLHUSK\TESTDLL1.CPP:19:#define AFXAPI_DATA __based(__segname("_DATA"))
MFC\SAMPLES\DLLHUSK\TESTDLL2.CPP:19:#define AFXAPI_DATA __based(__segname("_DATA"))
MFC\SRC\STDAFX.H:18:#define AFXAPI_DATA     __based(__segname("_DATA"))
MFC\SRC\STDAFX.H:19:#define AFXAPIEX_DATA   __based(__segname("_DATA"))
MFC\SRC\STDAFX.H:23:#define AFXAPI_DATA     __based(__segname("_DATA"))
MFC\SRC\STDAFX.H:26:#define AFXAPIEX_DATA   __based(__segname("_DATA"))
MFC\SRC\STDAFX.H:58:#define AFXAPP_DATA     __based(__segname("_DATA"))
```

**Reply by lightspot21** [reply]
From the given snippet these look like plain `#define`s? Unless I'm missing something, the parser should take care of them. (Not an MFC wiz though so maybe I'm missing something here)

**Reply by Wall-AF** [reply]
`_bcalloc` is just one of the function definitions.

**Reply by astrelsky** [reply]
`-D_based= ` problem solved

**Comment by shamefulCake1** [comment]
@lightspot21 did you have any luck with this project?

**Reply by lightspot21** [reply]
Hey! Unfortunately life happened and I couldn't find the time to work on this 😅 although I still need it

**Comment by playday3008** [comment]
https://github.com/NationalSecurityAgency/ghidra/discussions/2727#discussioncomment-16357451
